add Paramiko tests

Friz-zy opened the pull request:
Add minimal test cases for `ParamikoSSHVendor` according to https://github.com/dulwich/dulwich/issues/364

I spent whole day trying to do the adopt test ssh server and then I understanded, that we need test only paramiko api, not ssh connection itself xD

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,77 @@
+    def test_run_command_password(self):
+        vendor = ParamikoSSHVendor(allow_agent=False, look_for_keys=False)
+        try:
+            vendor.run_command(
```
Please use self.assertRaises when you're expecting an exception to be raised.

That's shorter, and that way, the test fails correctly if no exception is being raised - and you'll get a proper error message to that extend.

jelmer commented:
I think we do actually want to run an actual SSH server in process, since that's the only thing that will verify paramiko isn't changing underneath us.

Friz-zy commented:
1) it's much harder to implement
2) we don't test ssh connection in any SSH Vendor, it's not our competence
paramiko have it's own tests and specs, we should check only that we call with right arguments

jelmer commented:
On Fri, Mar 30, 2018 at 11:50:32AM -0700, Filipp Frizzy wrote:
> 1) it's much harder to implement
> 2) we don't test ssh connection in any SSH Vendor, it's not our competence
> paramiko have it's own tests and specs, we should check only that we call with right arguments
I agree it's much harder to implement. For the subprocess-based code,
I'm comfortable that e.g. ssh isn't going to change their
command-line.

However, I don't trust Paramiko to not break their APIs - I've seen
too many instances of them breaking in the past. This is why I mentioned
end-to-end tests in
https://github.com/dulwich/dulwich/issues/364#issuecomment-368324133

-- 
Jelmer Vernooĳ <jelmer@jelmer.uk>
PGP Key: https://www.jelmer.uk/D729A457.asc

Friz-zy commented:
1) I didn't see any problem with paramiko ssh implementation with other projects and we can't test each paramiko version about regressions anyway
2) mentioned in #364 problem is about functions api that now covered with this PR
I spent whole day and didn't do the right realisation of mosk ssh server...

jelmer commented:
On Fri, Mar 30, 2018 at 07:20:39PM +0000, Filipp Frizzy wrote:
> 1) I didn't see any problem with paramiko ssh implementation with other projects and we can't test each paramiko version about regressions anyway
Yes, that's another worry. What we've seen with bzr in the past is
that future versions of Paramiko would introduce a change in behaviour that
breaks past versions of the application that uses it.

An end-to-end test at least means that we'll find out from e.g. travis
that something is broken. A trivial test (like this one) can mean we
won't find out until there's a release out and users are actually
using the problematic combination of paramiko and dulwich. Your test
just verifies that Paramiko raises the right exception if it

Dulwich' testsuite already sets up a HTTP server to test its HTTP client
implementation against.

Paramiko ships its own SSH server implementation that we can use for
testing.

> 2) mentioned in #364 problem is about functions api that now covered with this PR
> I spent whole day and didn't do the right realisation of mosk ssh server...
I appreciate that this costs time, but I was pretty clear in the bug that
and end-to-end test was what I was after before moving the paramiko
support out of contrib.

For the SubprocessSSHVendor, we do actually verify that
subprocess.call() gets called with the right the ssh right arguments.
This change doesn't do the equivalent for paramiko just yet either.


-- 
Jelmer Vernooĳ <jelmer@jelmer.uk>
PGP Key: https://www.jelmer.uk/D729A457.asc

dzhuang commented:
I happened to test the vendor in another project, using `mock`, just to see if the exprected errors were raised by `paramiko`. I think that's is enough. I can push it here if you like that.

dzhuang commented:
Sorry, my idea is also in terms of unit test, instead of integration test (suggested by @jelmer ).

dzhuang commented:
FWIW, the integration test can use a [`NullServer`](https://github.com/paramiko/paramiko/blob/283cb267ac2ee2967f14bd9c78a1addb47c26343/tests/test_client.py#L58) used by `paramiko` to do the client test.

dzhuang commented:
(Here)[https://github.com/inducer/relate/pull/522/files#diff-b6ddd22d5858073729aa557842d3c9d3R211]'s the test I just did (but the project rewrote a `DulwichParamikoSSHVendor` class, for some minor customization).

dzhuang commented:
Should be https://github.com/dulwich/dulwich/pull/618

Friz-zy commented:
@dzhuang thank you for the link into paramiko client tests, I understand and fixed my realization :)

Friz-zy commented:
@jelmer finally works ^_^

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,163 @@
+# paramiko_vendor.py
```
test_paramiko_vendor.py

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,163 @@
+        super(Server, self).__init__(*args, **kwargs)
+
+    def check_channel_exec_request(self, channel, command):
+        COMMANDS.append(command)
```
Please make COMMANDS an instance variable, so state is not kept between tests.

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,163 @@
+        return True
+
+    def check_auth_password(self, username, password):
+        if (username == USER) and (password == PASSWORD):
```
parentheses are unnecessary here (and discouraged by PEP8)

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,163 @@
+
+    def check_auth_publickey(self, username, key):
+        pubkey = paramiko.RSAKey.from_private_key(StringIO(CLIENT_KEY))
+        if (username == USER) and (key == pubkey):
```
Same here

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,163 @@
+    def tearDown(self):
+        if hasattr(self, 'transport'):
+            self.transport.close()
+        if hasattr(self, 'socket'):
```
This shouldn't be conditional, since setUp() sets up the socket.

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,163 @@
+        self.thread.start()
+
+    def tearDown(self):
+        if hasattr(self, 'transport'):
```
Please use self.addCleanup(self.transport.close) where transport is created instead.

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,163 @@
+        self.socket.bind(('127.0.0.1', 0))
+        self.socket.listen(5)
+        self.port = self.socket.getsockname()[1]
+        self.thread = threading.Thread(target=self._run)
```
Where does the thread get terminated?

Friz-zy replied:
On self.transport.close call as I understand

jelmer replied:
self.transport doesn't have a handle to self.thread, who can it stop the thread?

Friz-zy replied:
Thread that started in `setUp` is finished when `_run` finished
`_run` wait for connection, accept it, create a paramiko transport from connection and then execute paramiko server. maybe this server also work in separate thread, maybe no - I didn't check paramiko internals. _run and self.thread should finished after this and `self.transport.start_server(server=server)` doesn't block it

jelmer replied:
Please make sure that the threads are terminated when the test finishes. Right now, this can easily leak a running SSH server in a thread.

jelmer replied:
Implemented in follow-up.

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,163 @@
+            '127.0.0.1', 'test_run_command_with_privkey',
+            username=USER, port=self.port, pkey=key)
+
+        self.assertTrue(b'test_run_command_with_privkey' in COMMANDS)
```
self.assertIn

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,163 @@
+        super(Server, self).__init__(*args, **kwargs)
+
+    def check_channel_exec_request(self, channel, command):
+        COMMANDS.append(command)
```
Can you also add some stdin/stdout interaction?

Friz-zy replied:
How?

jelmer replied:
I imagine this would involve taking some action on the channel object, e.g. reading in some data there and writing something back.

Friz-zy replied:
Then there should be some prepared data for each test.
I don't think that we need it as we should check only that ssh connection is work and we can pass the command into server

jelmer replied:
It doesn't have to be valid git requests; you can just pass in some sample strings (possibly via the constructor to Server?)

jelmer replied:
This doesn't test the transport path - it just sends a command to run, but doesn't verify that any actual data can go over the transport.

Friz-zy commented:
@jelmer I added socket timeout so any server thread would fail by timeout if tests would failed in the middle of process. Also I added test for stdout and fix for `_ParamikoWrapper` to handle it. For some reasons stderr doesn't work as I expected but I don't see that `read_stderr` is actually used in the code...

jelmer commented:
I did mean to actually kill the thread, not just set a timeout. With the timeout, it will still effectively leak (and live as long as the testsuite is running).

jelmer commented on dulwich/contrib/paramiko_vendor.py:
```diff
@@ -97,7 +97,7 @@ def read(self, n=None):
 
         # Closed socket
         if not data:
```
Why is this code being hit at all by the tests?

Friz-zy replied:
```
channel.send(b'stdout\n')
self.assertEqual(b'stdout\n', con.read(4096))
```

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,189 @@
+    def setUp(self):
+        self.commands = []
+        socket.setdefaulttimeout(10)
+        self.addCleanup(socket.setdefaulttimeout, None)
```
Why set the timeout on a socket that's going away?

jelmer commented on dulwich/contrib/test_paramiko_vendor.py:
```diff
@@ -0,0 +1,189 @@
+    def _run(self):
+        try:
+            conn, addr = self.socket.accept()
+        except:
```
Please catch specific exceptions. Why is this necessary?

Friz-zy commented:
Thread is not a subprocess, how you want to kill it?!
The `_run` thread is blocked by socket. Socket timeout would finish it if test failed before client connect to this socket

dzhuang commented:
Why not just follow the instruction at top of dulwich/contrib/test_paramiko_vendor.py (instantiate a sshclient) to do the test and make sure the vender and the wapper are used, asserting paramiko command or exceptions hehave as expected?

Friz-zy commented:
@dzhuang this way I did: new instance of wrapper connected into our virtual test ssh server in each test

jelmer commented on dulwich/contrib/paramiko_vendor.py:
```diff
@@ -97,7 +97,7 @@ def read(self, n=None):
         # Closed socket
         if not data:
-            return
+            return b''
```
Thanks, I've cherry-picked this change.

jelmer commented:
Merged.
